File: lib/ruby/signature/ast/members.rb
module Ruby
  module Signature
    module AST
      module Members
        class MethodDefinition
          attr_reader :name
          attr_reader :kind
          attr_reader :types
          attr_reader :annotations
          attr_reader :location

          def initialize(name:, kind:, types:, annotations:, location:)
            @name = name
            @kind = kind
            @types = types
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(MethodDefinition) &&
              other.name == name &&
              other.kind == kind &&
              other.types == types
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ kind.hash ^ types.hash
          end
        end

        module Var
          attr_reader :name
          attr_reader :type
          attr_reader :location

          def initialize(name:, type:, location:)
            @name = name
            @type = type
            @location = location
          end

          def ==(other)
            other.is_a?(self.class) && other.name == name && other.type == type
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type.hash
          end
        end

        class InstanceVariable
          include Var
        end

        class ClassInstanceVariable
          include Var
        end

        class ClassVariable
          include Var
        end

        module Mixin
          attr_reader :name
          attr_reader :args
          attr_reader :annotations
          attr_reader :location

          def initialize(name:, args:, annotations:, location:)
            @name = name
            @args = args
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(self.class) && other.name == name && other.args == args
          end

          def eql?(other)
            self == other
          end

          def hash
            self.class.hash ^ name.hash ^ args.hash
          end
        end

        class Include
          include Mixin
        end

        class Extend
          include Mixin
        end

        module Attribute
          attr_reader :name
          attr_reader :type
          attr_reader :ivar_name
          attr_reader :annotations
          attr_reader :location

          def initialize(name:, type:, ivar_name:, annotations:, location:)
            @name = name
            @type = type
            @ivar_name = ivar_name
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(self.class) &&
              other.name == name &&
              other.type == type &&
              other.ivar_name == ivar_name
          end

          alias eql? ==

          def hash
            self.class.hash ^ name.hash ^ type.hash ^ ivar_name.hash
          end
        end

        class AttrReader
          include Attribute
        end

        class AttrAccessor
          include Attribute
        end

        class AttrWriter
          include Attribute
        end

        module LocationOnly
          attr_reader :location

          def initialize(location:)
            @location = location
          end

          def ==(other)
            other.is_a?(self.class)
          end

          alias eql? ==

          def hash
            self.class.hash
          end
        end

        class Public
          include LocationOnly
        end

        class Private
          include LocationOnly
        end

        class Alias
          attr_reader :new_name
          attr_reader :old_name
          attr_reader :kind
          attr_reader :annotations
          attr_reader :location

          def initialize(new_name:, old_name:, kind:, annotations:, location:)
            @new_name = new_name
            @old_name = old_name
            @kind = kind
            @annotations = annotations
            @location = location
          end

          def ==(other)
            other.is_a?(self.class) &&
              other.new_name == new_name &&
              other.old_name == old_name &&
              other.kind == kind
          end

          alias eql? ==

          def hash
            self.class.hash ^ new_name.hash ^ old_name.hash ^ kind
          end
        end
      end
    end
  end
end

File: rust/ruby-rbs/src/ast/members.rs
use crate::ast::annotation::Annotation;
use crate::ast::comment::Comment;
use crate::ast::location::{
    AliasMemberLocation, AttributeMemberLocation, MethodDefinitionLocation, MixinMemberLocation,
    VariableMemberLocation,
};
use crate::ast::method_type::MethodType;
use crate::ast::types::Type;
use crate::ids::{SymbolId, TypeName};

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub enum Member {
    MethodDefinition(MethodDefinitionMember),
    InstanceVariable(InstanceVariableMember),
    ClassInstanceVariable(ClassInstanceVariableMember),
    ClassVariable(ClassVariableMember),
    Include(IncludeMember),
    Extend(ExtendMember),
    Prepend(PrependMember),
    AttrReader(AttrReaderMember),
    AttrWriter(AttrWriterMember),
    AttrAccessor(AttrAccessorMember),
    Public(PublicMember),
    Private(PrivateMember),
    Alias(AliasMember),
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum Visibility {
    Public,
    Private,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum MethodKind {
    Instance,
    Singleton,
    SingletonInstance,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct MethodDefinitionMember {
    pub name: SymbolId,
    pub kind: MethodKind,
    pub overloads: Vec<MethodDefinitionOverload>,
    pub annotations: Vec<Annotation>,
    pub location: Option<MethodDefinitionLocation>,
    pub comment: Option<Comment>,
    pub overloading: bool,
    pub visibility: Option<Visibility>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct MethodDefinitionOverload {
    pub method_type: MethodType,
    pub annotations: Vec<Annotation>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct InstanceVariableMember {
    pub name: SymbolId,
    pub ty: Type,
    pub location: Option<VariableMemberLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ClassInstanceVariableMember {
    pub name: SymbolId,
    pub ty: Type,
    pub location: Option<VariableMemberLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ClassVariableMember {
    pub name: SymbolId,
    pub ty: Type,
    pub location: Option<VariableMemberLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct IncludeMember {
    pub name: TypeName,
    pub args: Vec<Type>,
    pub annotations: Vec<Annotation>,
    pub location: Option<MixinMemberLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct ExtendMember {
    pub name: TypeName,
    pub args: Vec<Type>,
    pub annotations: Vec<Annotation>,
    pub location: Option<MixinMemberLocation>,
    pub comment: Option<Comment>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct PrependMember {
    pub name: TypeName,
    pub args: Vec<Type>,
    pub annotations: Vec<Annotation>,
    pub location: Option<MixinMemberLocation>,
    pub comment: Option<Comment>,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum AttributeKind {
    Instance,
    Singleton,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub enum IvarName {
    Unspecified,
    Empty,
    Name(SymbolId),
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct AttrReaderMember {
    pub name: SymbolId,
    pub ty: Type,
    pub ivar_name: IvarName,
    pub kind: AttributeKind,
    pub annotations: Vec<Annotation>,
    pub location: Option<AttributeMemberLocation>,
    pub comment: Option<Comment>,
    pub visibility: Option<Visibility>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct AttrAccessorMember {
    pub name: SymbolId,
    pub ty: Type,
    pub ivar_name: IvarName,
    pub kind: AttributeKind,
    pub annotations: Vec<Annotation>,
    pub location: Option<AttributeMemberLocation>,
    pub comment: Option<Comment>,
    pub visibility: Option<Visibility>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct AttrWriterMember {
    pub name: SymbolId,
    pub ty: Type,
    pub ivar_name: IvarName,
    pub kind: AttributeKind,
    pub annotations: Vec<Annotation>,
    pub location: Option<AttributeMemberLocation>,
    pub comment: Option<Comment>,
    pub visibility: Option<Visibility>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct PublicMember {
    pub location: Option<crate::ast::location::LocationRange>,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct PrivateMember {
    pub location: Option<crate::ast::location::LocationRange>,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub enum AliasKind {
    Instance,
    Singleton,
}

#[derive(Clone, Debug, Eq, PartialEq, Hash)]
pub struct AliasMember {
    pub new_name: SymbolId,
    pub old_name: SymbolId,
    pub kind: AliasKind,
    pub annotations: Vec<Annotation>,
    pub location: Option<AliasMemberLocation>,
    pub comment: Option<Comment>,
}
